grjobs/utils/supervised_utils.py

```diff
@@ -0,0 +1,177 @@
+            if len(green.split(' ')) > 1:
+                green_counts = job['clean_description'].count(green)
+                normalised_counts = green_counts/len(job_toks)
+                green_counts_feature.append((job['id'], normalised_counts))
+            if len(green.split(' ')) == 1:
+                green_toks_count = job_toks.count(green)
+                normalised_tok_counts = green_toks_count/len(job_toks)
+                green_counts_feature.append((job['id'], normalised_tok_counts))
+
+    keys = set(k for k, _ in green_counts_feature)
+    green_totals = {unique_key: sum(v for k, v in green_counts_feature if k == unique_key) for unique_key in keys}
+
+    with_green_counts = []
+    for job_ad in clean_labelled:
+        for green_key, green_total in green_totals.items():
+            if str(job_ad['id']) == str(green_key):
+                job_ad['green_count'] = green_total
+                with_green_counts.append(job_ad)
+    
+    return with_green_counts
+
+
+def tfidf_vectorise(labelled_clean_jobs): #this needs to be scaled up 
+    """
+    This function vectorises job descriptions and stacks the normalised
+    green count feature. 
+
+    :return: stacked job description and green count matrices
+    """
+    texts = [clean['clean_description'] for clean in labelled_clean_jobs]
+    green_counts = np.array([job['green_count'] for job in labelled_clean_jobs])
+
+    vectorizer = TfidfVectorizer(min_df = 0.05, max_df = 0.60)
+    features = vectorizer.fit_transform(texts).toarray()
+    
+    stacked_features = np.hstack((features, green_counts[:,None]))
+
+    labelled_transform = []
+    for job, vec in zip(labelled_clean_jobs, stacked_features):
+        job['vec'] = vec
+        labelled_transform.append(job)
+
+    return labelled_transform
+    
+def oversample_training_ads(labelled_transform):
+    """
+    This function uses SMOTE to oversample labelled 'green' jobs 
+    in an effort to address the imbalanced nature of the task.
+    
+    :return: a list of training matrices and training labels.
+    """
+    X_train = [job['vec'] for job in labelled_transform if 'label' in job.keys()]
+    y_train = [job['label'] for job in labelled_transform if 'label' in job.keys()]
+    
+    oversample = SMOTE()
+    X_over, y_over = oversample.fit_resample(X_train, y_train)
+    
+    return X_over, y_over
+
+def classify_jobs(X_train, y_train, labelled_transform):
+    """
+    This function trains an xgboost classifier on oversampled job matrices
+    and predicts whether a job is 'green' or not.
+    :return: a list of predictions 
+    """
+    
+    not_labeled_features = np.array([job['vec'] for job in labelled_transform if 'label' not in job.keys()])
+    
+    xgb = XGBClassifier(max_depth= 7, min_child_weight= 1)
+    xgb.fit(np.array(X_train), y_train)
+    predict = xgb.predict(not_labeled_features)
+    
+    return predict
+
+def get_green_jobs(limit, pretrained_model_path):
```

Comment:
Is the idea that this function runs the whole pipeline? If so, I think it would make more sense to factor this out into a script module or ideally to create a Metaflow flow